### ISSUE TITLE
Windows spawned after certain screen lockers wrongly drawn on top

### ISSUE THREAD

**Issue by TauPan** [qtile/qtile#676]
When I lock my screen with xscreensaver-command -lock and then send myself a notification, the notification will appear over the locked screen.

Note that I do not use the notification factilities provided with qtile but run my own notification daemon from ~/.xsession (xfce4-notifyd being started on demand by notify-send).

```
xscreensaver-command -lock; sleep 3; notify-send Hullo 'I am on top'                                                                                
```

This is also true for notifications from chromium, and might pose a mild security issue because of potential sensitive contents (i.e. from google mail), depending on user setup.

I've verified that the notification windows are _not_ override-redirect windows (if they were, it appears that the only fix would be to stop using those programs, as the window manager can't do anything about them).

We've discussed this on irc before:

``` #!irc
12:32       TauPan > Hm, I just noticed that I'm affected by this: http://www.jwz.org/xscreensaver/faq.html#popup-windows
12:48       TauPan > I've just checked, and the chrome notifications that appear over xscreensaver are *not* override-redirect windows, just _NET_WM_STATE(ATOM) = _NET_WM_STATE_SKIP_TASKBAR, 
                     _NET_WM_STATE_ABOVE
12:49       TauPan > So the fix proposed by jwz to map windows with XRestackWindows instead of XRaiseWindow might apply here.
12:49       TauPan > (but I didn't inspect qtile code yet.)
[...]
17:32        tych0@> TauPan: interesting, i didn't see either in our code, actually
[...]
09:42       TauPan > tych0: you mean, xtile uses neither XRestackWindows nor XRaiseWindow (or whatever their equivalents in xcb are)?
[...]
18:01        tych0@> TauPan: well, we don't call their xcb equivalents, we might stack the windows differently, though
18:01        tych0@> now that i think abou tit, we probably do
[...]
09:29       TauPan > tych0: it's easy to reproduce for me... I lock my screen with xscreensaver and then send myself a notification via Android: Do Button -> Pushbullet and the notification 
                     will appear on top of the screensaver. It's probably even easier with xscreensaver-command -lock; sleep 3; notify-send Hullo 'I am on top' 
09:30       TauPan > tych0: That works... I just verified it.
09:30       TauPan > tych0: (note that I do not use qtiles notification widget but run xfce4-notifyd
09:30       TauPan > )
[...]
16:50        tych0@> TauPan: yeah, i don't doubt that it's a bug :)
16:50        tych0@> TauPan: do you mind filing it?
16:52       TauPan > tych0: no problem, sorry, I was just lazy... will file it tomorrow
```

**Comment by ramnes** [comment]
I can confirm this, it's happening with ``i3lock`` also.

**Comment by jagguli** [comment]
Any movement on this ? this is pretty high priority for me, I use qtile at work and can have my personal messages popup over my locked screen, I could help fix and test this

**Comment by marcmerlin** [comment]
So, this is clearly not a new bug, but still present after 2 years.
Should I assume that there is no fix and no workaround?
My main issue is google-chrome writing notification windows on top of xscreensaver which is a privacy issue when those notifications are sensitive (like IMs) and visible to anyone in the office looking at my screensaver when I'm not around.

**Comment by ramnes** [comment]
You should try to use that violent `above` attribute that forces a window to be completely on the front of everything based on common screensavers X properties... unless we already use it for notifications, eh. But I don't think that's the case.

**Comment by ramnes** [comment]
Also:

> Should I assume that there is no fix and no workaround?

No, the only thing you should assume is that it wasn't important enough for people to spend time to find a fix for it. If it's important for you, then please give us a hand; you know how it works. :wink:

**Comment by marcmerlin** [comment]
@ramnes I think your answer confused me. Did you explain to me how to patch qtile so that it it's even more in front of everything?
My issue is that I don't want notifications to be in front of my xscreensaver, and they already are. I'm not sure if there is a good way to tell my window manager (enlightenment) to make sure xscreensaver is in front of all notifications. I'm thinking that the needed fix would be more in xscreensaver?

**Comment by ramnes** [comment]
Yep, you indeed look confused. :smile: 

What I'm proposing you here is to either:
- add a hook in your config that tries to detect for each new window if it's a screensaver and if so, place it above;
- or do the same inside Qtile directly.

**Comment by marcmerlin** [comment]
Gotcha, thanks for explaining.

**Comment by ramnes** [comment]
Does it still happen? I think I haven't seen this bug for quite a while.

**Comment by marcmerlin** [comment]
I haven't seen this for a while, but I've also switched to enlightenment's built in screensaver.

**Comment by TauPan** [comment]
I didn't see this either for quite a while now. I've switched to i3lock on all boxes now, but even before that (with xscreensaver) I think I didn't see it any more.

I just tried to reproduce this via:
~~~
i3lock; sleep 3; notify-send -t 100000 Hullo 'I am on top'
~~~

And I can only see the notification after I unlock the display, not before. Ok to close for me.

**Comment by tcld** [comment]
Does anybody have an idea how we could create a test for this type of scenario?

**Comment by m-col** [comment]
I looked through the source for slock and it doesn't look like there is anything special about a screensaver window, so I'm unsure if anything special would be needed in a test to replicate a lock. The notification window would need the `_NET_WM_WINDOW_TYPE_NOTIFICATION` window type. Then I guess you can just test their stacking order?

**Comment by tcld** [comment]
Thank you for the pointers. I assume a lockscreen/screensaver does at least have the fullscreen-flag set? Otherwise it would be expected behaviour that a notification can be display above it.

I haven't really done much with the tests yet, but I'll try my hands at this test.

**Comment by m-col** [comment]
Slock doesn't seem to set any flags. It seems to use XRaiseWindow upon start and
upon every received event, including when notification windows appear, and that
is what is making it appear above everything.

**Comment by ramnes** [comment]
Saw this again quite recently with i3lock.

**Comment by tcld** [comment]
From testing for #2029, I noticed this problem exists for *every* `_Window` spawned or raised after `i3lock` (I have not tested `xscreensaver`). However, other screenlockers are not necessarily affected (e.g. `slock`).
As @ramnes pointed out in the mentioned PR, the issue extends to restarts of Qtile (while i3lock is active), where e.g. `Bar`s are drawn on top of the lock screen. He mentioned, too, that the issue can be reproduced from the commandline with `i3lock & qtile cmd-obj -o cmd -f restart` - just substitute your screenlocker and your qtile-path.

Maybe this issue should be renamed a little? Something like "Windows spawned/raised after certain screenlockers wrongly drawn on top" might encompass the situation better.

**Comment by ramnes** [comment]
Yeah, no need for a restart actually, any window will be drawn on top. `i3lock & xterm` demonstrates the problem just as well. It probably means that we just don't pay attention to whatever X11 mechanism i3lock and others use to stay on top.

**Comment by tcld** [comment]
The code can be found [here](https://github.com/i3/i3lock/blob/master/xcb.c#L126), maybe somebody with more X11-knowledge than me is willing to take a look. m-col looked at the code for `dmenu` yesterday and said that it checked whether anything was overlapping and, if so, raised itself. I assume that's what `slock` is doing, too, which would explain why it doesn't suffer from this issue.

With only `StackMode.Above` and `StackMode.Below` available for layering windows (there might be more that I don't know of), it looks to me as if Qtile would have to manage *all* windows to properly solve this issue. Or am I overlooking something?

**Comment by ramnes** [comment]
Fixed by 3ce1a785ee81b8f8b4486391f9a92e19a2de70d2
